torch/csrc/distributed/c10d/nvshmem_extension.cu

```diff
@@ -0,0 +1,390 @@
+           ++split_idx) {
+        if (nvshmem_uint64_test(&split_signals[split_idx], NVSHMEM_CMP_EQ, 2)) {
+          const int split_begin = split_idx * split_size;
+          const int chunk_begin = split_begin + blockIdx.x * chunk_size;
+          nvshmem_int_put_signal_nbi(
+              output_ptr + chunk_begin,
+              output_ptr + chunk_begin,
+              chunk_size,
+              &split_signals[split_idx],
+              1,
+              NVSHMEM_SIGNAL_SET,
+              next_global_rank);
+          split_signals[split_idx] = 0;
+          forwarded += 1;
+        }
+      }
+
+      if (forwarded == world_size - 1 && received == world_size - 1) {
+        nvshmem_quiet();
+        break;
+      }
+    }
+  } else {
+    // ===============
+    // Reduction wraps
+    // ===============
+    const int thread_idx = threadIdx.x - warp_size;
+    const int num_threads = blockDim.x - warp_size;
+
+    for (int step = 0; step < world_size - 1; ++step) {
+      if (thread_idx == 0) {
+        int split_idx = ring_buf_pop(&acc_queue);
+        acc_split_idx = split_idx;
+      }
+      asm volatile("bar.sync 0, 512;" : : : "memory");
+
+      const int split_idx = acc_split_idx;
+      const size_t split_begin = split_idx * split_size;
+      const size_t chunk_begin = split_begin + blockIdx.x * chunk_size;
+
+      for (size_t offset = chunk_begin;
+           offset < std::min(chunk_begin + chunk_size, numel);
+           offset += num_threads) {
+        if (offset + thread_idx < numel) {
+          output_ptr[offset + thread_idx] =
+              output_ptr[offset + thread_idx] + input_ptr[offset + thread_idx];
+        }
+      }
+      asm volatile("bar.sync 0, 512;" : : : "memory");
+
+      if (thread_idx == 0) {
+        if (split_idx != rank) {
+          split_signals[split_idx] = 2;
+        }
+      }
+    }
+  }
+}
+
+at::Tensor nvshmem_reduce_scatter_out(
+    at::Tensor& input,
+    std::string group_name,
+    at::Tensor& out) {
+  auto input_hdl = c10d::symmetric_memory::rendezvous(input, group_name);
+  auto out_hdl = c10d::symmetric_memory::rendezvous(out, group_name);
+  int rank = input_hdl->get_rank();
+  int world_size = input_hdl->get_world_size();
+  auto team = group_to_team(group_name, input_hdl->get_rank_to_global_rank());
+
+  void* input_ptr = input_hdl->get_buffer_ptrs()[rank];
+  void* output_ptr = out_hdl->get_buffer_ptrs()[rank];
+  size_t numel = input.numel();
+  void* signal_pad_ptr = input_hdl->get_signal_pad_ptrs()[rank];
+  void* rank_to_global_rank = input_hdl->get_rank_to_global_rank_dev();
+  void* args[] = {
+      &input_ptr,
+      &output_ptr,
+      &numel,
+      &signal_pad_ptr,
+      &rank,
+      &world_size,
+      &rank_to_global_rank,
+      &team};
+
+  dim3 grid_dim(32), block_dim(544);
+  auto stream = at::cuda::getCurrentCUDAStream();
+  nvshmemx_barrier_on_stream(team, stream);
+  nvshmemx_collective_launch(
```

Comment:
Just for my learning, we need to use `nvshmemx_collective_launch` because `nvshmem_all_reduce_kernel` uses some nvshmem APIs. Is this correct?

Reply:
From documentation:
> The nvshmemx_collective_launch function must be used to launch CUDA kernels on the GPU when the CUDA kernels use NVSHMEM synchronization or collective APIs (e.g., nvshmem_wait, nvshmem_barrier, nvshmem_barrier_all, or any other collective operation). CUDA kernels that do not use synchronizing NVSHMEM APIs (or that do not use NVSHMEM APIs at all), are not required to be launched by this API. This call is collective across the PEs in the NVSHMEM job.